src/error.rs

```diff
@@ -1,43 +1,47 @@
 use std::{fmt::Display, io::ErrorKind, path::PathBuf};
 
-/// Error type for this program.
+/// An error type that can hold a `PathBuf`` as an optional context to an `io::Error`.
 #[derive(Debug)]
-pub enum Error {
-    /// File not found.
-    NotFound(PathBuf),
-    Other(String),
+pub struct Error {
+    path: Option<PathBuf>,
+    inner: std::io::Error,
 }
-use Error::*;
 
-pub fn from_io_error(path: PathBuf) -> impl Fn(std::io::Error) -> Error {
-    move |e: std::io::Error| -> Error {
-        match e.kind() {
-            ErrorKind::NotFound => NotFound(path.clone()),
-            _ => Other(e.to_string()),
-        }
+impl Error {
+    pub fn is_not_found(&self) -> bool {
+        self.inner.kind() == ErrorKind::NotFound && self.path.is_some()
     }
 }
 
-pub fn from_walkdir_error(path: PathBuf) -> impl Fn(walkdir::Error) -> Error {
-    move |e: walkdir::Error| -> Error {
-        match e.io_error() {
-            Some(_) => NotFound(path.clone()),
-            None => Other(e.to_string()),
+impl<E: Into<std::io::Error>> From<E> for Error {
```

Comment:
Yeah, nice. I only discovered the `Into` trait after writing this. Quite handy and also quite magical.